mempool/src/core_mempool/index.rs

```diff
@@ -256,33 +257,81 @@ impl TimelineIndex {
 /// (because their sequence number is too high)
 /// we keep separate index to be able to efficiently evict them when Mempool is full
 pub struct ParkingLotIndex {
-    data: BTreeSet<TxnPointer>,
+    // DS invariants:
+    // 1. for each entry (account, txns) in `data`, `txns` is never empty
+    // 2. for all accounts, data.get(account_indices.get(`account`)) == (account, sequence numbers of account's txns)
+    data: Vec<(AccountAddress, BTreeSet<u64>)>,
+    account_indices: HashMap<AccountAddress, usize>,
+    size: usize,
 }
 
 impl ParkingLotIndex {
     pub(crate) fn new() -> Self {
         Self {
-            data: BTreeSet::new(),
+            data: vec![],
+            account_indices: HashMap::new(),
+            size: 0,
         }
     }
 
     /// add transaction to index
     pub(crate) fn insert(&mut self, txn: &MempoolTransaction) {
-        self.data.insert(TxnPointer::from(txn));
+        let sender = &txn.txn.sender();
+        let sequence_number = txn.txn.sequence_number();
+        let is_new_entry = match self.account_indices.get(sender) {
+            Some(index) => {
+                if let Some((_account, seq_nums)) = self.data.get_mut(*index) {
+                    seq_nums.insert(sequence_number)
+                } else {
+                    unreachable!("parking lot invariant violated: for account {}, account index exists but missing entry in data", sender);
+                }
+            }
+            None => {
+                let seq_nums = [sequence_number].iter().cloned().collect::<BTreeSet<_>>();
+                self.data.push((*sender, seq_nums));
+                self.account_indices.insert(*sender, self.data.len() - 1);
+                true
+            }
+        };
+        if is_new_entry {
+            self.size += 1;
+        }
     }
 
     /// remove transaction from index
     pub(crate) fn remove(&mut self, txn: &MempoolTransaction) {
-        self.data.remove(&TxnPointer::from(txn));
+        let sender = &txn.txn.sender();
+        if let Some(index) = self.account_indices.get(sender).cloned() {
+            if let Some((_account, txns)) = self.data.get_mut(index) {
+                if txns.remove(&txn.txn.sequence_number()) {
+                    self.size -= 1;
+                }
+
+                // maintain DS invariant
+                if txns.is_empty() {
+                    // remove account with no more txns
+                    self.data.swap_remove(index);
+                    self.account_indices.remove(sender);
+
+                    // update DS for account that was swapped in `swap_remove`
+                    if let Some((swapped_account, _)) = self.data.get(index) {
+                        self.account_indices.insert(*swapped_account, index);
```

Comment:
Doesn't this introduce a possible concurrency bug where this thread is swapping an element but the account_indices map hasn't been updated yet and another thread tries to access that element?